include/kernel.h

```diff
@@ -3312,6 +3332,161 @@ extern int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
  */
 extern int k_delayed_work_cancel(struct k_delayed_work *work);
 
+/**
+ * @brief Initialize a delayable work item.
+ *
+ * This routine initializes a delayable work item prior to its first
+ * use.
+ *
+ * @warning This function must not be invoked on a delayable work item that
+ * has been passed to k_work_schedule_for_queue() unless it is known that the
+ * work is no longer delayed, pending on the work queue, or being processed by
+ * the work queue.  This state cannot be inferred from k_work API return
+ * values.
+ *
+ * @param work Address of delayable work item.
+ * @param handler Function to invoke each time work item is processed.
+ *
+ * @return N/A
+ */
+extern void k_work_delayable_init(struct k_work_delayable *work,
+				  k_work_handler_t handler);
+
+/**
+ * @brief Get the parent delayable work structure from a work pointer.
+ *
+ * This function is necessary when a @c k_work_handler_t function is passed to
+ * k_work_schedule_for_queue() and the handler needs to access data associated
+ * with the work item.
+ *
+ * @param work Address passed to the work handler
+ *
+ * @return Address of the containing @c k_delayable_work structure.
+ */
+static inline struct k_work_delayable *
+k_work_delayable_from_work(struct k_work *work)
+{
+	return CONTAINER_OF(work, struct k_work_delayable, work);
+}
+
+/**
+ * @brief Schedule a delayable work item to be submitted to a queue.
+ *
+ * This routine schedules work item @p work to be processed by workqueue @p
+ * work_q after @p delay.  Only when the timeout completes is the work item
+ * actually submitted to the workqueue and becomes pending.
+ *
+ * If @p delay is not @c K_NO_WAIT then the work item will be scheduled to be
+ * submitted after @p delay.
+ *
+ * If @p delay is @c K_NO_WAIT then the effect is the same as invoking
+ * k_work_submit_to_queue().
+ *
+ * In either case any previously scheduled incomplete delay is cancelled.
+ *
+ * @note If a previous delay timeout had completed or was cancelled, and the
```

Comment:
I think I don't entirely understand the part "or was cancelled". I thought that the work item would not be scheduled if the previous delay had been successfully cancelled.

Reply:
Yes, that seems to be imprecise.  I'll tweak the language.  The point is that regardless of what happens in this call if the work item is pending it can be processed twice after this call is made.